### DISCUSSION TITLE
Heap size does not change after function call.

### DISCUSSION THREAD

**Discussion by paniabhisek** [bloomberg/memray#568] [Q&A] [answered]
Changes I did to  `memray/docs/example/fibonacci/fib.py`
1. `n` to `10_000` due to segmentation fault.
2. Added `sleep(2)` after `run()`

Next commands:

    PYTHONMALLOC=malloc memray run --trace-python-allocators fib.py
    memray flamegraph --temporal memray-fib.py.<id>.bin

But I do not see any deallocation of heap size. What is the reason?

This is also shown as memory leak if I give `--leaks` option

**Comment by pablogsal** [comment]
I converted this issue into a Q&A discussion. Please do not open issues for general questions if you are not sure they are bugs. This helps us a lot maintaining the repo. Thanks for understanding 🙏

**Comment by pablogsal** [comment]
This is happening because CPython doesn't release that memory at teardown because it's still on the freelists of the interpreter, if you force the memory to be allocated to be higher you will find that once you surpass the freelist limit that memory it's indeed released.

**Comment by paniabhisek** [comment]
How do I ensure memray to show leaks correctly then ?

**Reply by pablogsal** [reply]
Memray shows leaks correctly if you pass —leaks to the flame graph command

**Comment by godlygeek** [comment, marked as answer]
> This is happening because CPython doesn't release that memory at teardown because it's still on the freelists of the interpreter

No, it's nothing so complicated as that...

> I do not see any deallocation of heap size.

When I perform the steps you've given, I do see the heap size drop:
![Screenshot 2024-03-14 2 04 36 AM](https://github.com/bloomberg/memray/assets/29423/cca0f833-9a68-4a0e-b712-a56a3c15879c)

That circle shows the spot where `fib1()` has returned (and its temporary list has been freed) but `fib2()` has not yet built up a large cache. As the program continues running, the cache owned by `fib2()` gets larger and larger, until eventually it's even larger than the temporary list that was created while `fib1()` ran. That point marks the program's heap high water mark. 2 seconds later the main script finishes, after the sleep completes. Memray then correctly tells you that the cache owned by fib2 has been leaked. You told Memray to run `fib.py`, and while `fib.py` was running a large dict was created, and when `fib.py` finished running that large dict still existed. That's a leak, as far as Memray is concerned. That cache was allocated while the script ran and was not freed before the script finished.

You may be thrown off by the fact that `cache` is a [mutable default argument](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments). That cache is shared across all calls to `fib2`, and so it's not temporary, and continues to exist even after `fib2` has returned.

**Reply by paniabhisek** [reply]
I forgot about the mutable default argument. Very helpful!!
